kiali/kiali-ui
# KIALI-2641 Add Wizard for 3scale adapter config

This PRs refers to JIRA
https://issues.jboss.org/browse/KIALI-2641
and it needs backend PR https://github.com/kiali/kiali/pull/1050 with the API services.

Summary
----
Maistra will include a new Istio 3Scale adapter [1] that will allow to autenticate Services deployed in Istio with a 3Scale API. 

The configuration of this integration is somehow complex as a user needs to create Istio objects like rules, handlers and template instances.

The goal of this PR is to provide a high level integration from the UI to make this setup easier.

For this first step, we have defined a new Wizard at Service Details level. 
This Wizard only will be visible when the 3scale adapter is present in the cluster, otherwise, this PR will be transparent to the rest of the users.

![image](https://user-images.githubusercontent.com/1662329/57390377-7dceb580-71bc-11e9-8133-361e58423165.png)

When the 3scale adapter is available user will be able to "Add a 3Scale" rule for a service or to "Delete" an existing one. Note that this operation is 1to1, so, Kiali will only permit one 3Scale Rule per Service.

This will means that a specific service will query 3Scale for authentication, following the described policies defined in [1].

If cluster has not defined any 3Scale Handler, first action a user will do is to create one. To perform this operation, user needs to supply some parameters provided by the 3Scale Admin site.

![image](https://user-images.githubusercontent.com/1662329/57390661-2f6de680-71bd-11e9-9a39-3785add94290.png)

Once this handler is created; the handler is selected for the service that will be linked.

![image](https://user-images.githubusercontent.com/1662329/57390781-6643fc80-71bd-11e9-8678-4a8e2a640990.png)

Note, that user can create multiple handlers, but it will be typically only one per cluster, that will be used for multiple services. But multiple handlers can be created using the "Show Advanced Options".

![image](https://user-images.githubusercontent.com/1662329/57391006-d9e60980-71bd-11e9-88ac-db06708c4317.png)

Basic delete/update operations on handler(s) are supported from the Wizard:

![image](https://user-images.githubusercontent.com/1662329/57391068-fc782280-71bd-11e9-9ec8-e820204c1f0a.png)

Once the 3Scale Rule is created, a particular service is "linked" with 3Scale API using and indicated into the Service Description details:

![image](https://user-images.githubusercontent.com/1662329/57391123-27627680-71be-11e9-97d9-7654a7ab6afc.png)

Note that in the future, there can be more complex operations, like to perform a Rule on mulitple Services in a single operation. 

That is out of the scope of this PR.

How to test
---

There are several preliminary steps before you can test this feature.

First is to use install Maistra enabling the 3Scale adapter. To do that, you need to use a special custom installation yaml like this one:

```
apiVersion: "istio.openshift.com/v1alpha1"
kind: "Installation"
metadata:
  name: "istio-installation"
  namespace: istio-operator
spec:
  threeScale:
    enabled: true
```
That will deploy the 3scale adapter under the istio-system namespace:
![image](https://user-images.githubusercontent.com/1662329/57391450-e028b580-71be-11e9-9277-e5420de43d5c.png)

Also, 3Scale adapter needs to enable policies in Istio, if not, Rules won't take any effect without showing any warning/error on logs (yeap, I hope this would save you some time):

```
echo "Check policies...."
oc -n istio-system get cm istio -o jsonpath="{@.data.mesh}" | grep disablePolicyChecks
cd $ISTIO_HOME
echo "Updating policies..."
helm template install/kubernetes/helm/istio --namespace=istio-system -x templates/configmap.yaml --set global.disablePolicyChecks=false | oc -n istio-system replace -f -
```

For these tests I've created a special API in 3Scale called Kiali, under the API I prepared some configuration to be used with a special demo app called travel-agency:

https://github.com/lucasponce/travel-comparison-demo

So, first you need to deploy the travel-agency app using

```
https://github.com/lucasponce/travel-comparison-demo/blob/master/deploy-travel-agency.sh

Note: no need to install the full example, just the travel-agency is enough for this 3scale testing
```

Next, you would need to expose the travels service with a gateway

```
https://github.com/lucasponce/travel-comparison-demo/blob/master/travel_agency_gateway.yaml
```

And you can invoke the travel service through the gateway using something like this script

```
#!/bin/bash

export INGRESS_HOST=$(oc get po -l istio=ingressgateway -n istio-system -o 'jsonpath={.items[0].status.hostIP}')
export INGRESS_PORT=$(oc -n istio-system get service istio-ingressgateway -o jsonpath='{.spec.ports[?(@.name=="http2")].nodePort}')

TRAVEL_AGENCY_HOST="Host:www.travel-agency.com"
TRAVELS_SERVICE="http://${INGRESS_HOST}:${INGRESS_PORT}"

while true
do
    for city in paris rome london
    do
        for user in vip normal
        do
            RESULT=$(curl -H "${TRAVEL_AGENCY_HOST}" -H "user:${user}" -o /dev/null -s -w %{http_code} ${TRAVELS_SERVICE}/travels/${city}?user_key=4406b917c719df23055919c134651ee4)
            echo "Travel quota for ${city} from ${user} => $RESULT"
            sleep 1
        done
        RESULT=$(curl -H "${TRAVEL_AGENCY_HOST}" -H "user:${user}" -o /dev/null -s -w %{http_code} ${TRAVELS_SERVICE}/travels/${city})
        echo "Travel quota for ${city} from anonymous => $RESULT"
        sleep 1
    done
done
```

Note that this script will add a special 3scale query parameters "user_key" to propagate the user key of the demo user connected with 3Scale (you need to use this value, as it's a valid code).

Then, you can use the following 3scale parameters to create handlers/rule in the Wizard:

```
{
  "name": "my-test-handler",
  "serviceId": "2555417792901",
  "systemUrl": "https://kiali-admin.3scale.net",
  "accessToken": "15586d0814cf49c5fd8fef46940e774063476c3845a368d218ac6edc6befb054"
}

Note: "name" can be anything, but serviceId, systemUrl and accessToken are valid
```

How to validate things are working as expected
---

Once that you have created a 3Scale Rule under the "travels" service, only the calls with the valid user_key are responding a 200 Ok, the others should have a 403 Error.
When no rule is created, all calls should return a 200 Ok.

```
[...]
Travel quota for london from vip => 200
Travel quota for london from normal => 200
Travel quota for london from anonymous => 403
[...]
```

Also, in Istio Config, we can check that the Wizard generates a pair of Hanlder/Instances resources per "3Scale Handler" and a specific "Istio Rule" per link between "Service and 3Scale".

[1] https://github.com/3scale/3scale-istio-adapter
